perl/lib/Sanger/CGP/Vaf/Process/Variant.pm

```diff
@@ -600,73 +603,109 @@ Inputs
 
 sub _fetch_reads {
 my ($self,$sam_object,$region,$Reads_FH)=@_;
-
-my ($mate_info,%mapped_length);
-my $read_counter=0;
-        $sam_object->fetch($region, sub {
-        my $a = shift;
+    my $mate_info;
+    my $read_counter=1;
+    my $callback = sub {
+        my ($a, $mq) = @_;
         # \& bitwise comparison
         ##Ignore read if it matches the following flags:
         #Brass-ReadSelection.pm
-        return if($self->{'_mq'} && ($a->qual <= $self->{'_mq'}) );
-
+        return if(defined $mq && ($a->qual <= $mq) );
         return if $a->flag & $Sanger::CGP::Vaf::VafConstants::DEFAULT_READS_EXCLUDE_FETCH_MATE;
-
-        my $mseqid = $a->mate_seq_id;
-        my $seqid = $a->seq_id;
         #target gives read seq as it comes from sequencing machine i.e softclipped bases included
         my $qseq = $a->target->dna();
         return if(index(uc($qseq), 'N') != -1);
-        my $name=$a->display_name;
-        #my $strand = $a->strand;
-        my $mstart = $a->mate_start;
-        my $start = $a->start;
-        $read_counter++;
-        print  $Reads_FH ">$name\_$read_counter\n$qseq\n";
-    # fetch mate only if on another chromosome
-        if(defined $mseqid and defined $seqid and ($seqid ne $mseqid)) {
-        #if(defined $mseqid and defined $seqid ) {
-            $mate_info->{$name}="$mseqid:$mstart-$mstart";
+
+        my $mtid = $a->mtid;
+        my $tid = $a->tid;
+        my $name=$a->qname;
+        my $mpos = $a->mpos;
+        printf $Reads_FH ">%s\_%d\n%s\n", $name, $read_counter++, $qseq;
+        # fetch mate only if on another chromosome
+        if(defined $mtid && defined $tid && ($tid ne $mtid)) {
+            push @{$mate_info->{$mtid}->{$mpos}}, $name;
         }
+    };
+    #$sam_object->fetch($region, $callback);
+    $sam_object->hts_index->fetch(
+        $sam_object->hts_file,
+        $sam_object->header->parse_region($region),
+        $callback,
+        $self->{'_mq'}
+    );
+
+    my $mate_sets = $self->_collapse_mate_locs($mate_info);
+    for my $mate_set(@{$mate_sets}) {
+        $self->_fetch_mate_seq($sam_object, $mate_set, $Reads_FH);
+    }
+}
+
+=head2 _collapse_mate_locs
+
+Collapses reads into ranges to reduce calls to fetch
+
+=cut
+
+sub _collapse_mate_locs {
+    my ($self, $mate_info) = @_;
+    my @read_sets;
 
-    });
-    #added separately as it was pulling only one read
-    foreach my $key (keys %$mate_info)
-    {
-        $self->_fetch_mate_seq($sam_object,$mate_info->{$key},$key,$Reads_FH);
+    for my $seqid(keys %{$mate_info}) {
+        for my $mstart(sort {$a<=>$b} keys %{$mate_info->{$seqid}} ) {
+            my %mnames = map {$_ => 1 } @{$mate_info->{$seqid}->{$mstart}};
+            if(@read_sets == 0 || $read_sets[-1]->{'seqid'} ne $seqid) {
+                push @read_sets, {'seqid' => $seqid, 'low' => $mstart, 'high' => $mstart, 'reads' => \%mnames};
+                next;
+            }
+            if($mstart - $MATE_BUFFER <= $read_sets[-1]->{'high'}) {
```

Comment:
Nothing more than a sensible range to group many calls into a single fetch range instead of doing hundreds against the same compressed block of the BAM/CRAM.